File: chat/src/main/java/com/qiscus/sdk/ui/view/LayoutConfiguration.java
/*
 * Copyright (c) 2016 Qiscus.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.qiscus.sdk.ui.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.Gravity;

import com.qiscus.sdk.R;

class LayoutConfiguration {
    private int orientation = QiscusFlowLayout.HORIZONTAL;
    private boolean debugDraw = false;
    private float weightDefault = 0;
    private int gravity = Gravity.LEFT | Gravity.TOP;
    private int layoutDirection = QiscusFlowLayout.LAYOUT_DIRECTION_LTR;

    public LayoutConfiguration(Context context, AttributeSet attributeSet) {
        TypedArray a = context.obtainStyledAttributes(attributeSet, R.styleable.QiscusFlowLayout);
        try {
            this.setOrientation(a.getInteger(R.styleable.QiscusFlowLayout_android_orientation, QiscusFlowLayout.HORIZONTAL));
            this.setDebugDraw(a.getBoolean(R.styleable.QiscusFlowLayout_debugDraw, false));
            this.setWeightDefault(a.getFloat(R.styleable.QiscusFlowLayout_weightDefault, 0.0f));
            this.setGravity(a.getInteger(R.styleable.QiscusFlowLayout_android_gravity, Gravity.NO_GRAVITY));
            this.setLayoutDirection(a.getInteger(R.styleable.QiscusFlowLayout_layoutDirection, QiscusFlowLayout.LAYOUT_DIRECTION_LTR));
        } finally {
            a.recycle();
        }
    }

    public int getOrientation() {
        return this.orientation;
    }

    public void setOrientation(int orientation) {
        if (orientation == QiscusFlowLayout.VERTICAL) {
            this.orientation = orientation;
        } else {
            this.orientation = QiscusFlowLayout.HORIZONTAL;
        }
    }

    public boolean isDebugDraw() {
        return this.debugDraw;
    }

    public void setDebugDraw(boolean debugDraw) {
        this.debugDraw = debugDraw;
    }

    public float getWeightDefault() {
        return this.weightDefault;
    }

    public void setWeightDefault(float weightDefault) {
        this.weightDefault = Math.max(0, weightDefault);
    }

    public int getGravity() {
        return this.gravity;
    }

    public void setGravity(int gravity) {
        this.gravity = gravity;
    }

    public int getLayoutDirection() {
        return layoutDirection;
    }

    public void setLayoutDirection(int layoutDirection) {
        if (layoutDirection == QiscusFlowLayout.LAYOUT_DIRECTION_RTL) {
            this.layoutDirection = layoutDirection;
        } else {
            this.layoutDirection = QiscusFlowLayout.LAYOUT_DIRECTION_LTR;
        }
    }
}

File: android/src/main/java/com/qiscus/sdk/chat/presentation/android/view/LayoutConfiguration.kt
package com.qiscus.sdk.chat.presentation.android.view

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import com.qiscus.sdk.chat.presentation.android.R

internal class LayoutConfiguration(context: Context, attributeSet: AttributeSet) {

    var orientation = FlowLayout.HORIZONTAL
        set(orientation) = if (orientation == FlowLayout.VERTICAL) {
            field = orientation
        } else {
            field = FlowLayout.HORIZONTAL
        }

    var isDebugDraw = false

    var weightDefault = 0f
        set(weightDefault) {
            field = Math.max(0f, weightDefault)
        }

    var gravity = Gravity.LEFT or Gravity.TOP

    var layoutDirection = FlowLayout.LAYOUT_DIRECTION_LTR
        set(layoutDirection) = if (layoutDirection == FlowLayout.LAYOUT_DIRECTION_RTL) {
            field = layoutDirection
        } else {
            field = FlowLayout.LAYOUT_DIRECTION_LTR
        }

    init {
        val a = context.obtainStyledAttributes(attributeSet, R.styleable.FlowLayout)
        try {
            this.orientation = a.getInteger(R.styleable.FlowLayout_android_orientation, FlowLayout.HORIZONTAL)
            this.isDebugDraw = a.getBoolean(R.styleable.FlowLayout_debugDraw, false)
            this.weightDefault = a.getFloat(R.styleable.FlowLayout_weightDefault, 0.0f)
            this.gravity = a.getInteger(R.styleable.FlowLayout_android_gravity, Gravity.NO_GRAVITY)
            this.layoutDirection = a.getInteger(R.styleable.FlowLayout_layoutDirection, FlowLayout.LAYOUT_DIRECTION_LTR)
        } finally {
            a.recycle()
        }
    }
}